### ISSUE TITLE
g:set cannot set page scope variables in self-referencing template

### ISSUE THREAD

**Issue by danduke** [grails/grails-core#9238]
In a brand new grails project using 2.3.9 or 2.5.1 (have not tested other versions), I cannot set a page scoped variable's value using g:set the first time the variable is seen on a page, if the variable has been set before on another page.  The previously set value is used.  However, if I print the value first, then set it, it works.

This feels like a premature optimization during page compilation, but I have not dug deeply enough to say for sure.

This is a simple enough example that I am going to include the code inline; there were no other changes to my project after I ran "create-app".

**index.gsp**
```
<!DOCTYPE html>
<html>
<head>
</head>
<body>

<h1>INDEX</h1>

SETTING TO 1: <g:set var="var1" value="1" />
IS: ${var1}, SET WORKED: ${var1 == "1"}<br/>

SETTING TO 2: <g:set var="var1" value="2" />
IS: ${var1}, SET WORKED: ${var1 == "2"}<br/>

<g:render template="/testPage1" model="${[level: 0]}" />

</body>
</html>
```

**_testPage1.gsp**
```
<%@ page contentType="text/html;charset=UTF-8" %>
<html>
<head>
  <title></title>
</head>

<body>

<h1>SUB PAGE LEVEL ${level}</h1>

%{-- If the next line is uncommented, then everything works. --}%
%{--Started at: ${var1}<br/>--}%

SETTING TO 3: <g:set var="var1" value="3" />
IS: ${var1}, SET WORKED: ${var1 == "3"}<br/>

SETTING TO 4: <g:set var="var1" value="4" />
IS: ${var1}, SET WORKED: ${var1 == "4"}<br/>

<g:if test="${level < 2}">
  <g:render template="/testPage1" model="${[level: level + 1]}" />
</g:if>
</body>
</html>
```

**Result (minus formatting)**
```
INDEX
SETTING TO 1: IS: 1, SET WORKED: true
SETTING TO 2: IS: 2, SET WORKED: true
SUB PAGE LEVEL 0
SETTING TO 3: IS: 3, SET WORKED: true
SETTING TO 4: IS: 4, SET WORKED: true
SUB PAGE LEVEL 1
SETTING TO 3: IS: 4, SET WORKED: false
SETTING TO 4: IS: 4, SET WORKED: true
SUB PAGE LEVEL 2
SETTING TO 3: IS: 4, SET WORKED: false
SETTING TO 4: IS: 4, SET WORKED: true
```

I expect all "SET WORKED" lines to show "true".

**Comment by graemerocher** [comment]
Grails 2.x is in maintenance mode and only blocking bugs and bugs reported by OCi support customers are being fixed. Please upgrade to Grails 3.x.

**Comment by danduke** [comment]
I realize this is quite delayed, but in case anyone else stumbles across this and needs to know: confirmed fixed in 3.2.6.
